site/sfguides/src/production_ready_rag_with_trulens_snowflake_cortex/production_ready_rag_trulens_snowflake_cortex.md

```diff
@@ -0,0 +1,483 @@
+  "Can you recommend some resources for learning Streamlit?",
+  "What are some common use cases for Streamlit?",
+  "How can I deploy a streamlit app on the cloud?",
+  "How do I add a logo to streamlit?",
+  "What is the best way to deploy a Streamlit app?",
+  "How should I use a streamlit toggle?",
+  "How do I add new pages to my streamlit app?",
+  "How do I write a dataframe to display in my dashboard?",
+  "Can I plot a map in streamlit? If so, how?",
+  "How do vector stores enable efficient similarity search?",
+  "How do I prevent my child from using the internet?",
+  "What should I pack for a camping trip this weekend?",
+  "How do I defend myself against bear attacks?"
+]
+```
+
+```python
+with tru_rag as recording:
+  for prompt in prompts:
+    rag.query(prompt)
+
+tru.get_leaderboard()
+```
+
+## Use Guardrails
+Duration: 7
+
+In addition to making informed iteration, we can also directly use feedback results as guardrails at inference time. In particular, here we show how to use the context relevance score as a guardrail to filter out irrelevant context before it gets passed to the LLM. This both reduces hallucination and improves efficiency.
+
+![Context Filter Guardrails](./assets/guardrail_context_filtering.png)
+
+To do so, we'll rebuild our RAG using the `@context-filter decorator` on the method we want to filter, and pass in the feedback function and threshold to use for guardrailing.
+
+```python
+# note: feedback function used for guardrail must only return a score, not also reasons
+f_context_relevance_score = (
+  Feedback(provider.context_relevance, name = "Context Relevance")
+)
+
+from trulens_eval.guardrails.base import context_filter
+
+# note: feedback function used for guardrail must only return a score, not also reasons
+f_context_relevance_score = (
+  Feedback(provider.context_relevance, name = "Context Relevance")
+  .on_input()
+  .on(Select.RecordCalls.retrieve.rets)
+)
+
+from trulens_eval.guardrails.base import context_filter
+
+class filtered_RAG_from_scratch:
+
+  def __init__(self):
+    self.retriever = CortexSearchRetriever(session=session, limit_to_retrieve=4)
+  @instrument
+  @context_filter(f_context_relevance_score, 0.75, keyword_for_prompt="query")
+  def retrieve_context(self, query: str) -> list:
+    """
+    Retrieve relevant text from vector store.
+    """
+    results = self.retriever.retrieve(query)
+    return results
+
+  @instrument
+  def generate_completion(self, query: str, context_str: list) -> str:
+    """
+    Generate answer from context.
+    """
+    completion = Complete("mistral-large",query)
+    return completion
+
+  @instrument
+  def query(self, query: str) -> str:
+    context_str = self.retrieve_context(query=query)
+    completion = self.generate_completion(query=query, context_str=context_str)
+    return completion
+
+filtered_rag = filtered_RAG_from_scratch()
+```
+
+## Conclusion And Resources
+Duration: 1
+
+### What You Learned
```

Comment:
The entire doc you were talking like "we" but now it's "you". I think it works better as "we".